Make grad point to bucket buffer in DDP to save memory usage

zhaojuanmao opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #43282 add doc regarding that grads are pointing to bucket views in DDP
* **#41954 Make grad point to bucket buffer in DDP to save memory usage**

Make both variable.grad() and grad in distautograd context point to bucket buffer in DDP to save memory usage.
In this case, grad will be view of bucket buffer tensors, in order to make it compatiable with optimizer.zero_grad(), we
made changes in https://github.com/pytorch/pytorch/pull/41283.

Also be noted that we can not make variable.grad() pointing to bucket buffer during construction time, because we want to
keep grad undefined for unused parameters.


Test Plans:
For roberta_base model with ~1GB parameters, peak memory dropped ~1GB (8250MB-7183MB).  Per iteration latency (0.982s ->0.909s), 8% speed up; will rerun a few times to confirm the speed up
For resnet model with ~97M parameters, peak memory dropped ~100MB (3089MB -> 2988MB). Per iteration latency has no change (0.122s -> 0.123s)

Differential Revision: [D22707857](https://our.internmc.facebook.com/intern/diff/D22707857/)

zhaojuanmao committed:
Make grad point to bucket buffer in DDP to save memory usage

Make both variable.grad() and grad in distautograd context point to bucket buffer in DDP to save memory usage.
In this case, grad will be view of bucket buffer tensors, in order to make it compatiable with optimizer.zero_grad(), we
made changes in https://github.com/pytorch/pytorch/pull/41283.

Also be noted that we can not make variable.grad() pointing to bucket buffer during construction time, because we want to
keep grad undefined for unused parameters.

Differential Revision: [D22707857](https://our.internmc.facebook.com/intern/diff/D22707857/)

[ghstack-poisoned]

zhaojuanmao committed:
Update on "Make grad point to bucket buffer in DDP to save memory usage"


Make both variable.grad() and grad in distautograd context point to bucket buffer in DDP to save memory usage.
In this case, grad will be view of bucket buffer tensors, in order to make it compatiable with optimizer.zero_grad(), we
made changes in https://github.com/pytorch/pytorch/pull/41283.

Also be noted that we can not make variable.grad() pointing to bucket buffer during construction time, because we want to
keep grad undefined for unused parameters.

Differential Revision: [D22707857](https://our.internmc.facebook.com/intern/diff/D22707857/)

[ghstack-poisoned]

zhaojuanmao committed:
Update on "Make grad point to bucket buffer in DDP to save memory usage"


Make both variable.grad() and grad in distautograd context point to bucket buffer in DDP to save memory usage.
In this case, grad will be view of bucket buffer tensors, in order to make it compatiable with optimizer.zero_grad(), we
made changes in https://github.com/pytorch/pytorch/pull/41283.

Also be noted that we can not make variable.grad() pointing to bucket buffer during construction time, because we want to
keep grad undefined for unused parameters.

Differential Revision: [D22707857](https://our.internmc.facebook.com/intern/diff/D22707857/)

[ghstack-poisoned]

zhaojuanmao committed:
Update on "Make grad point to bucket buffer in DDP to save memory usage"


Make both variable.grad() and grad in distautograd context point to bucket buffer in DDP to save memory usage.
In this case, grad will be view of bucket buffer tensors, in order to make it compatiable with optimizer.zero_grad(), we
made changes in https://github.com/pytorch/pytorch/pull/41283.

Also be noted that we can not make variable.grad() pointing to bucket buffer during construction time, because we want to
keep grad undefined for unused parameters.

Differential Revision: [D22707857](https://our.internmc.facebook.com/intern/diff/D22707857/)

[ghstack-poisoned]

pritamdamania87 commented on torch/csrc/distributed/c10d/reducer.cpp:
```diff
@@ -751,24 +758,47 @@ void Reducer::initialize_buckets(
+      bucket_view = contents.narrow(0, offset, length).view(v.sizes());
     }
+    replica.bucket_views.push_back(bucket_view);
+    // If grad has already been defined/calculated in previous iterations,
+    // and the grad does not point to bucket_view, that means bucket_view
+    // is newly allocated, if this function is called inside initialize_bucket,
+    // old grad needs to be copied into new bucket_view for initialization
+    // and let grad point to bucket_view;
+    // otherwise this function is called for communication hook, bucket_view
+    // has the updated results in new tensor, just let grad point to
+    // bucket_view. If grad is not defined, do not let it point to bucket_view.
+    // E.g., for global_unused parameters, their grads should be kept as being
+    // undefined.
```
This change is complicated enough that we probably want to add a NOTE in a single place which describes the overall logic in detail and then we can reference that note here.

zhaojuanmao replied:
sounds good! will add notes

pritamdamania87 commented on torch/csrc/distributed/c10d/reducer.cpp:
```diff
@@ -751,24 +758,47 @@ void Reducer::initialize_buckets(
+    // E.g., for global_unused parameters, their grads should be kept as being
+    // undefined.
+    runGradCallbackForVariable(v, [&](auto& grad) {
+      if (grad.defined() && !grad.is_alias_of(bucket_view)) {
```
I'm wondering if my understanding is correct about what happens in the first iteration:

1. In the first iteration, the grads are not defined when we enter here via `initialize_buckets`. 
2. As a result, the grad has separate storage and the bucket has separate storage.
3. When we execute the logic in `mark_variable_ready_dense`, the grad produced by backward is copied into the bucket and now the grad points to the bucket instead of having its own storage.
4. Now, from the second iteration onward it seems like we'll have only one copy?

Does this mean we'll still have two copies of the gradient very briefly for each parameter? I guess this should be fine in most cases, but in some cases a parameter could be pretty large (large nn.EmbeddingBag). 

> Also be noted that we can not make variable.grad() pointing to bucket buffer during construction time, because we want to
keep grad undefined for unused parameters.

I was wondering if we can just set the gradients of the unused parameters as undefined at the end of the backward pass to achieve this? At the beginning of the next forward pass, we can make these grads point back to the bucket.

zhaojuanmao replied:
for 4. second iteration will call initialize_bucket again to rebuild buckets. if users did not zero out grad or set grad to be none in their user codes, we need to copy old bucket to new bucket in this case, and maintain two copies for the bucket, when parameter is large, the memory saving is not large. I'm trying to add codes where I will check if grad is zero tensor, I will free the grad before allocating a new bucket. it will work well for most cases when users usually zero out grads. 

"I was wondering if we can just set the gradients of the unused parameters as undefined at the end of the backward pass to achieve this? At the beginning of the next forward pass, we can make these grads point back to the bucket."
==> this sounds like a good idea! possibly it can make codes cleaner, but how to set gradient to be undefined?

zhaojuanmao replied:
Had second thought. I think we can not set gradients of the unused parameters as undefined at the end of the backward pass, because if the parameter is unused, we just want to keep the grad not updated, not keep the grad undefined.

pritamdamania87 replied:
I'm wondering if we can just have a flag that indicates whether the parameter was undefined or not at the beginning  of the backward pass? The process would be something like this:

1. For the first iteration, if grads are not undefined, we copy them to the bucket and make the grads point to the bucket. For undefined grads, we just point them to a bucket with all zeros. For each parameter we store a flag of whether or not they were undefined at the beginning of the backward pass.
2. Once the backward pass is complete, if a parameter was undefined at the beginning of the backward pass and it was unused, we set its grad to None.
3. For the next backward pass, we repeat the same process where we identify the undefined parameters, point them back to the bucket and keep a flag if they were undefined at the beginning of the backward pass.

zhaojuanmao replied:
sounds reasonable, I think your suggestion can help further avoid copy if grad is none before backward starts in some iterations. The number of grads usually is hundreds, so extra memory usage is trivial as well.

zhaojuanmao replied:
Had thought about it again:). The above suggestion assumes grads will be mutated in backward pass in place.  ===> but it is not guaranteed.   

if grads are mutated out of place in backward, above suggestion will not give correct results. 

although current diff (let grad point to bucket_view when grad is ready) can not save memory when grads are mutated out of place, it can still give correct results.

How do you think? I feel it is more safe to let grad point to bucket_view during backward in this way

pritamdamania87 replied:
Yes, you are right it won't work in the case where the user might modify the grads themselves. I think this PR looks good to me and I'll let @mrshenli stamp it.

Although, can we add some unit tests to verify everything works as expected with this PR when the user modifies `.grad` field themselves? These tests would ensure our logic for copying the grads in this case work as expected.

zhaojuanmao replied:
sounds good, let me add one more test for this case

zhaojuanmao replied:
actually I checked our unit tests that have already manually set grad to be none after iteration, this will trigger copying the grads here. and it is working as expected

pritamdamania87 commented:
Would be nice if we could also share some memory saving results as part of the PR description.

zhaojuanmao commented on torch/csrc/distributed/c10d/reducer.cpp:
```diff
@@ -311,52 +311,60 @@ void Reducer::mark_variable_ready_dense(VariableIndex index) {
-            bucket_view.sizes(),
-            ", strides() = ",
-            bucket_view.strides());
+      // Copy grad to bucket view buffer if grad and bucket_view are pointing
+      // to different storages, and then let grad point to bucket_view
+      // for saving memory and avoiding copies in subsquent iterations.
+      // In most cases, the copy is needed only at first
+      // iteration, there will be no copies in subsquent iterations.
+      // In rare cases, if users explicitly set grad to be none after every
+      // iteration, then it needs to copy grad to bucket_view in every
+      // iteration.
```
it could happen when users explicitly set grad to be none after optimizer.step()  in their codes

zhaojuanmao commented on torch/csrc/distributed/c10d/reducer.cpp:
```diff
@@ -660,7 +668,7 @@ void Reducer::initialize_buckets(
               variable_index < replicas_[replica_index].size(),
               "Out of range variable index specified.");
-          const auto& variable = replicas_[replica_index][variable_index];
+          auto& variable = replicas_[replica_index][variable_index];
```
forgot to revert it back, I was trying to make its grad point to an empty tensor, so that we can free the old bucket before allocating new bucket.  This does not work if users did not zero out grad or set grad to be none in their user codes, I'm still working on this part locally

zhaojuanmao committed:
Update on "Make grad point to bucket buffer in DDP to save memory usage"


Make both variable.grad() and grad in distautograd context point to bucket buffer in DDP to save memory usage.
In this case, grad will be view of bucket buffer tensors, in order to make it compatiable with optimizer.zero_grad(), we
made changes in https://github.com/pytorch/pytorch/pull/41283.

Also be noted that we can not make variable.grad() pointing to bucket buffer during construction time, because we want to
keep grad undefined for unused parameters.

Differential Revision: [D22707857](https://our.internmc.facebook.com/intern/diff/D22707857/)

[ghstack-poisoned]

mrshenli commented on torch/csrc/autograd/VariableTypeManual.cpp:
```diff
@@ -269,7 +269,12 @@ Tensor & detach_(Tensor & self) {
-      AT_ERROR("Can't detach views in-place. Use detach() instead");
+      AT_ERROR("Can't detach views in-place. Use detach() instead."
+               "If you are using DistributedDataParallel API for "
+               "training, we've made grads be views of allReduce "
+               "buffers, so you can not detach_ the grads directly."
+               "Fixing it by referring to how zero_grad is defined "
+               "inside optimizer.");
```
nit:

If you are using DistributedDataParallel (DDP) for training, gradients are views of DDP buckets, and hence `detach_()` cannot be called on these gradients. To fix this error, please refer to the `Optimizer.zero_grad()` function in `torch/optim/optimizer.py` as the solution.

mrshenli commented on torch/nn/parallel/distributed.py:
```diff
@@ -563,6 +563,12 @@ def forward(self, *inputs, **kwargs):
+        # inside rebuildBuckets.
+        # To save peak memory usage, call rebuildBuckets before the peak
+        # memory usage increases during computing outputs.
+        self.reducer.rebuildBuckets()
```
Is it true that call this in the first fired-fired autograd hook is already too late?

zhaojuanmao replied:
autograd hook will still record indices of grads in dynamic order, rebuildBuckets here is to use those recorded indices to build new buckets, so it should be fine to call it here

mrshenli commented on torch/nn/parallel/distributed.py:
```diff
@@ -563,6 +563,12 @@ def forward(self, *inputs, **kwargs):
+        # inside rebuildBuckets.
+        # To save peak memory usage, call rebuildBuckets before the peak
+        # memory usage increases during computing outputs.
+        self.reducer.rebuildBuckets()
```
Since the launching of this function is moved to Python, I wonder if the conditioning should also be moved to Python (I mean `has_rebuilt_bucket_`)? Otherwise, this looks like a per-iteration overhead.

zhaojuanmao replied:
I feel has_rebuilt_bucket_ should be kept in reducer, I can change the rebuildBucket name to prepare_forward() for better clarity

mrshenli commented on torch/csrc/distributed/c10d/reducer.cpp:
```diff
@@ -751,24 +758,49 @@ void Reducer::initialize_buckets(
+      bucket_view = contents.narrow(0, offset, length).view(v.sizes());
     }
+    replica.bucket_views.push_back(bucket_view);
+    // There are two cases to handle:
```
three cases?

zhaojuanmao commented on test/distributed/test_distributed.py:
```diff
@@ -1967,7 +1967,10 @@ def test_DistributedDataParallel_non_default_stream(self):
                 grad = net.module.weight.grad
                 if grad is not None:
-                    grad.detach_()
+                    if grad.grad_fn is not None:
```
sounds good

pritamdamania87 commented on torch/nn/parallel/distributed.py:
```diff
@@ -563,6 +563,12 @@ def forward(self, *inputs, **kwargs):
         if self.require_forward_param_sync:
             self._sync_params()
 
+        # It may allocate new buckets before deallocating old buckets
+        # inside rebuildBuckets.
+        # To save peak memory usage, call rebuildBuckets before the peak
+        # memory usage increases during computing outputs.
```
Maybe I'm missing something here, but in `initialize_buckets` it seems like we first clear the existing buckets before allocating new ones? (https://github.com/pytorch/pytorch/blob/master/torch/csrc/distributed/c10d/reducer.cpp#L624).

zhaojuanmao replied:
right, it cleared existing buckets without this diff. With this diff, variable.grad is pointing to the bucket tensor as well, so clear buckets_ will not clear the existing buckets

albanD commented on test/distributed/test_distributed.py:
```diff
@@ -1967,7 +1967,10 @@ def test_DistributedDataParallel_non_default_stream(self):
                 # Clear gradients manually
```
Why is this cleared manually and not using `net.module.zero_grad()` ? That would simplify this logic.

zhaojuanmao replied:
sounds good

albanD commented on torch/csrc/distributed/c10d/reducer.cpp:
```diff
@@ -311,52 +311,60 @@ void Reducer::mark_variable_ready_dense(VariableIndex index) {
+      // for saving memory and avoiding copies in subsquent iterations.
+      // In most cases, the copy is needed only at first
+      // iteration, there will be no copies in subsquent iterations.
+      // In rare cases, if users explicitly set grad to be none after every
```
nit: None

albanD reviewed (commented):
Looks mostly good.
I would add a comment here as well: https://github.com/pytorch/pytorch/blob/c660d2a9aea49d0fd2d2ce952b5c2aec3278d853/torch/csrc/autograd/functions/accumulate_grad.h#L176-L177  as now the DDP relies on the fact that AccumulateGrad always change the .grad implace when it exists (when no double backward) even if the variable and the .grad don't have the same layout!

zhaojuanmao commented:
@albanD  thanks for your review. I will add comments in accumulate_grad.h. But be noted, if grads are mutated in place, DDP can save memory; if grads are mutated out of place somehow, although DDP can not save memory, it is still working at it is today (checking grad is alias of bucket buffer or not, if not, copying grad to bucket buffer)

zhaojuanmao committed:
Update on "Make grad point to bucket buffer in DDP to save memory usage"


Make both variable.grad() and grad in distautograd context point to bucket buffer in DDP to save memory usage.
In this case, grad will be view of bucket buffer tensors, in order to make it compatiable with optimizer.zero_grad(), we
made changes in https://github.com/pytorch/pytorch/pull/41283.

Also be noted that we can not make variable.grad() pointing to bucket buffer during construction time, because we want to
keep grad undefined for unused parameters.


Test Plans:
For roberta_base model with 1GB parameters, this diff can save 1GB memory.
DDP without this diff, peak allocated memory during training loop is 8250 MB;
DDP with this diff, peak allocated memory during training loop is 7182 MB;

Differential Revision: [D22707857](https://our.internmc.facebook.com/intern/diff/D22707857/)

[ghstack-poisoned]

albanD commented:
Yes it will still give the right result I agree but it will defeat the purpose of this optimization (silently!). So it might be hard to detect that changes there actually break this optimization. Hence my request to add a comment there to make people aware that this can happen.

zhaojuanmao committed:
Update on "Make grad point to bucket buffer in DDP to save memory usage"


Make both variable.grad() and grad in distautograd context point to bucket buffer in DDP to save memory usage.
In this case, grad will be view of bucket buffer tensors, in order to make it compatiable with optimizer.zero_grad(), we
made changes in https://github.com/pytorch/pytorch/pull/41283.

Also be noted that we can not make variable.grad() pointing to bucket buffer during construction time, because we want to
keep grad undefined for unused parameters.


Test Plans:
For roberta_base model with 1GB parameters, this diff can save 1GB memory.
DDP without this diff, peak allocated memory during training loop is 8250 MB;
DDP with this diff, peak allocated memory during training loop is 7182 MB;

Differential Revision: [D22707857](https://our.internmc.facebook.com/intern/diff/D22707857/)

[ghstack-poisoned]

zhaojuanmao commented:
@mrshenli and @pritamdamania87 this diff is ready to take another look now, I also updated the test plans, both roberta and resnets got expected peak memory usage reduce, also there is 8% speed up for roberta model (I will rerun a few times to make sure it is not due to variance)

albanD commented on test/distributed/test_distributed.py:
```diff
@@ -2113,7 +2113,7 @@ def test_DistributedDataParallel_non_default_stream(self):
                 grad = net.module.weight.grad
                 if grad is not None:
-                    grad.detach_()
+                    grad.requires_grad_(False)
```
This is not equivalent as it won't clear the grad_fn if one exists.
It may be ok for this test though.
Is there a reason why you don't use the `zero_grad()` function from the Module?

zhaojuanmao replied:
I guess the author just wanted to set one grad in this module, not all grads in the module, so trying to follow the original style in this test

albanD commented on torch/csrc/autograd/functions/accumulate_grad.h:
```diff
@@ -161,6 +161,10 @@ struct TORCH_API AccumulateGrad : public Node {
         // valid operation which adds `new_grad` to `variable_grad` in
         // place. `variable_grad` is thus still referring to the same tensor
         // after the operation.
+        // Also DistributedDataParallel(DDP) package relies on grad is
```
grad is -> grad being

albanD commented on torch/csrc/autograd/functions/accumulate_grad.h:
```diff
@@ -161,6 +161,10 @@ struct TORCH_API AccumulateGrad : public Node {
+        // Also DistributedDataParallel(DDP) package relies on grad is
+        // mutated in place for saving peak memory usage. DDP will still
+        // work correctly if it is mutated out of place here, but it will
+        // increase one copy of grads peak memory usage.
```
Not sure this is clear. Do you mean that it will lead to one extra copy (and so slow down) and one extra Tensor to be alive at the same time (and so more memory)?

zhaojuanmao replied:
one extra tensor and more memory, I will refactor the comment to make it clear

albanD reviewed (commented):
The high level change looks ok to me. Even I didn't go into detail of all the code.

One thing that I think is missing is the user facing doc for this and the BC-breaking doc:
We should have in the DDP doc a clear explanation of what we're doing. So that users that try to save/inspect these grads actually expect to get views.
And we should mark this as BC breaking as it changes the behavior (in particular for people doing .detach_()) on these parameters. And you should add in the PR description a note of what exactly is changing and how user can get around the new errors they might encounter.

zhaojuanmao commented:
@albanD sounds good! will add explanation in DDP and BC breaking docs.

Also reran benchmarks a few times, there is ~7-8% speedup with this diff for roberta bert model consistently!

mrshenli commented on torch/csrc/distributed/c10d/reducer.cpp:
```diff
@@ -763,27 +773,57 @@ void Reducer::initialize_buckets(
 // (see Note:  "Gradient Layout Contract" in initialize_buckets).
 void Reducer::initialize_bucketviews(
```
bucketviews -> bucket_views?

and ditto for comments below

zhaojuanmao commented on torch/csrc/distributed/c10d/reducer.cpp:
```diff
@@ -64,7 +64,9 @@ Reducer::Reducer(
 
   // Initialize variable bucketing.
   // This can be reinitialized later after capturing runtime information.
+  std::unique_lock<std::mutex> lock(this->mutex_);
```
lock was inside initializeBuckets() before, I am moving this lock out of it here. So that we do not need to unlock before calling intializeBuckets() inside rebuildBuckets() (the unlock may result in race with other threads)

mrshenli replied:
hmm, which thread might lead to race here? I don't clearly recall the original reason for adding lock to `initialize_buckets`, was it because it was used in `mark_variable_ready`? After we move it to Reducer ctor here, is this lock still necessary?

zhaojuanmao replied:
initialize_buckets has lock originally even when it is called in reducer constructor only. I think the lock is added possibly because initialize_buckets was exposed to be reducer's public function, so it could be called by multiple threads and thus lock is needed in this public function. 

I agree with you that the lock may be redundant in constructor. But I'm still keeping the lock here to keep the same logic as before, just in case I missed something

mrshenli commented on torch/csrc/distributed/c10d/reducer.h:
```diff
@@ -193,7 +204,10 @@ class Reducer {
+  void initialize_bucketviews(
+      BucketReplica& replica,
+      at::Tensor& contents,
+      bool copy_to_bucket_view);
```
shall we add a comment here to explain what does `copy_to_bucket_view` mean and when it is used?

mrshenli commented on torch/csrc/distributed/c10d/reducer.h:
```diff
@@ -135,7 +146,7 @@ class Reducer {
   // For find_unused_parameters = false case, buckets are only rebuilt once,
   // the performance cost is negligible.
-  std::vector<std::vector<size_t>> rebuildBuckets();
+  void rebuildBuckets();
```
nit: I noticed that all other methods are using snake names, shall we use `rebuild_buckets` as name here as well?

mrshenli commented on torch/csrc/distributed/c10d/reducer.cpp:
```diff
@@ -763,27 +773,57 @@ void Reducer::initialize_buckets(
+    // 1. initialize_bucketviews could be called inside communication hook,
+    // bucket_view has the updated results in new tensor, just let grad point to
+    // bucket_view.
+    // 2. initialize_bucketviews could be called inside initialize_bucket when
```
initialize_bucket -> initialize_buckets

mrshenli commented on torch/csrc/distributed/c10d/reducer.cpp:
```diff
@@ -763,27 +773,57 @@ void Reducer::initialize_buckets(
+    // rebuildBuckets, if grad has already been defined/calculated in previous
+    // iteration, old grad needs to be copied into new bucket_view
+    // and let grad point to the new bucket_view;
+    // 3. initialize_bucketviews could be called inside initialize_bucket
```
ditto

mrshenli commented on torch/csrc/distributed/c10d/reducer.cpp:
```diff
@@ -763,27 +773,57 @@ void Reducer::initialize_buckets(
+    // 3. initialize_bucketviews could be called inside initialize_bucket
+    // during construction. When grad is not defined, do not let it point to
+    // bucket_view, because grads should be kept as being undefined for globally
+    // unused parameters.
```
Are 2 and 3 actually the same? they are both `copy_to_bucket_view = True`?

zhaojuanmao replied:
yes, both of them pass copy_to_bucket_view = True, but during construction time mostly grads are not defined. I am adding more comments to clarify it

mrshenli commented on torch/csrc/distributed/c10d/reducer.cpp:
```diff
@@ -763,27 +773,57 @@ void Reducer::initialize_buckets(
 }
 
+void Reducer::prepare_forward() {
+  std::lock_guard<std::mutex> lock(mutex_);
```
what race we are trying to protect here? I think this is the same discussion as in the Reducer ctor. I might miss sth, but it doesn't seem like there will be different thread mutating states here?

zhaojuanmao replied:
prepare_forward is exposed to be public function of reducer, generally it needs to protect variables if multiple threads call this function? also, it is possible multiple threads calls forwards(), right?

this is similar to prepare_for_backward() where lock is added as well

mrshenli replied:
> prepare_forward is exposed to be public function of reducer, generally it needs to protect variables if multiple threads call this function? also, it is possible multiple threads calls forwards(), right?

I see, yep, the lock is also there for `prepare_for_backward`.

mrshenli commented:
@zhaojuanmao all my comments are nits, except the two regarding locks. Will stamp when we have a conclusion on that.

mrshenli reviewed (approved):
Stamping to unblock. We can try to remove the lock in ctor in a followup PR.

zhaojuanmao committed:
Update on "Make grad point to bucket buffer in DDP to save memory usage"


Make both variable.grad() and grad in distautograd context point to bucket buffer in DDP to save memory usage.
In this case, grad will be view of bucket buffer tensors, in order to make it compatiable with optimizer.zero_grad(), we
made changes in https://github.com/pytorch/pytorch/pull/41283.

Also be noted that we can not make variable.grad() pointing to bucket buffer during construction time, because we want to
keep grad undefined for unused parameters.


Test Plans:
For roberta_base model with ~1GB parameters, peak memory dropped ~1GB (8250MB-7183MB).  Per iteration latency (0.982s ->0.909s), 8% speed up; will rerun a few times to confirm the speed up
For resnet model with ~97M parameters, peak memory dropped ~100MB (3089MB -> 2988MB). Per iteration latency has no change (0.122s -> 0.123s)

Differential Revision: [D22707857](https://our.internmc.facebook.com/intern/diff/D22707857/)

[ghstack-poisoned]

pritamdamania87 commented:
> For roberta_base model with ~1GB parameters, peak memory dropped ~1GB (8250MB-7183MB). Per iteration latency (0.982s ->0.909s), 8% speed up; will rerun a few times to confirm the speed up
For resnet model with ~97M parameters, peak memory dropped ~100MB (3089MB -> 2988MB). Per iteration latency has no change (0.122s -> 0.123s)

Had a general question about the results, it seems like we see 13% memory reduction in case of roberta and only 3% in case of resnet, do we know what is the reason for this? Is it also possible to share the break down of the 7183MB and 2988MB respectively to understand how much memory is used by each component of PyTorch?

It looks like we use optim.SGD in our benchmarks which has a momentum buffer of the same size as the parameter. So if we have X parameters, the total memory would be X (parameters) + X (optimizer state) + X (gradients) + X (DDP bucket) + activations (say A) = 4X + A. It seems like if we completely get rid of DDP bucket memory, we should see a memory reduction of about 20-25% (assuming activations are not very large). It would be good to understand where the additional memory is used.

zhaojuanmao commented:
The goal of this diff is to reduce the extra copy of grads in buckets, roberta_base model has ～1GB parameters, so this diff reduced around 1GB memory; resnet model has ~100MB parameters, so this diff reduced around 100MB memory; Now, ddp training peak memory is the same as local training peak memory. 

Regarding how breakdowns look like in the DDP training or local training, it is kind of out of scope for this diff. But it is quite interesting to explore in the future though
